File: 06_pgm_Avion/src/co/edu/campusucc/poo/Avion.java
package co.edu.campusucc.poo;

public abstract class Avion {
    public Avion() {
    }

    // propiedades
    private String marca;
    private String tipo;
    private int velocidad;

    public String toString() {
        return "Clase Abstracta--> Avion: " + getMarca() + " " + getTipo() + " ";
    }

    // Abstract Behavior
    public abstract void startingEngine();

    public abstract void stopingEngine();

    // Definir comportamiento
    public void speedUp() {
        String speedUpString = "SpeedUp in progres: ";
        for (int i = 0; i < velocidad; i++) {
            speedUpString += "🛫";
        }

        for (int i = 0; i < 10; i++) {
            speedUpString += "🛫";
            System.out.println(speedUpString);
            try {
                Thread.sleep(1000 - (i * 50));
            } catch (InterruptedException e) {
                System.err.println("⛔:" + e);
            }
        }
        addSpeed(10);

    }

    protected int addSpeed(int velocidad) {
        setVelocidad(this.velocidad + velocidad);
        return getVelocidad();
    }

    private int downSpeed(int velocidad) {
        setVelocidad(this.velocidad - velocidad);
        return getVelocidad();
    }

    public void currentSpeed() {
        System.out.println("Current Speed 🛫: " + getVelocidad());
    }

    public void landing() {
        for (int i = 0; i < 5; i++) {
            System.out.println("🛬🛬 Landing in Progress 🚷🚷");
            try {
                Thread.sleep(1000 - (i * 50));
            } catch (InterruptedException e) {
                System.err.println("⛔:" + e);
            }
        }

    }

    public void shootMissiles(){
        for (int i = 0; i < 5; i++) {
            System.out.println("🚀🚀 Missiles ready to fire 🔥🔥🔥");
            try {
                Thread.sleep(1000 - (i * 50));
            } catch (InterruptedException e) {
                System.err.println("⛔:" + e);
            }
        }
    }

    public void irrigation(){
        for (int i = 0; i < 5; i++) {
            System.out.println("🛩️🛩️ irrigation in process 💧💧💧");
            try {
                Thread.sleep(1000 - (i * 50));
            } catch (InterruptedException e) {
                System.err.println("⛔:" + e);
            }
        }
    }
    
    public void descent(){
        String speedUpString = "";
        for (int i = 0; i < velocidad; i++) {
            speedUpString += "🛬";
        }
        for (int i = 0; i < 10; i++) {
            speedUpString = speedUpString.substring(0, speedUpString.length() - 1);
            System.out.println("🛬🛬Descent in Progress🛬🛬: " + speedUpString);
            try {
                Thread.sleep(1000 - (velocidad + i * 50));
            } catch (InterruptedException e) {
                System.err.println("⛔:" + e);
            }
        }
        downSpeed(10);
    }

    // Getters and Setters
    public String getMarca() {
        return marca;
    }

    public void setMarca(String marca) {
        this.marca = marca;
    }

    public String getTipo() {
        return tipo;
    }

    public void setTipo(String tipo) {
        this.tipo = tipo;
    }
    public int getVelocidad() {
        return velocidad;
    }

    public void setVelocidad(int velocidad) {
        this.velocidad = velocidad;
    }


}

File: python/java_to_python/09_pgm_avion_interface/src/co/edu/campusucc/poo/abst/Avion.py
from abc import ABC, abstractmethod
import time
from src.co.edu.campusucc.poo.interfaces import IAvionStandar


class Avion(ABC, IAvionStandar):
    def __init__(self):
        self.marca = None
        self.tipo = None
        self.velocidad = 0

    def __str__(self):
        return f"Clase Abstracta--> Avion: {self.marca} {self.tipo} "

    # Abstract Behavior
    @abstractmethod
    def startingEngine(self):
        pass

    @abstractmethod
    def stopingEngine(self):
        pass

    # Definir comportamiento
    def speedUp(self):
        speedUpString = "SpeedUp in progres: "

        for i in range(self.velocidad):
            speedUpString += "🛫"

        for i in range(10):
            speedUpString += "🛫"
            print(speedUpString)

            try:
                time.sleep(1 - (i * 50))
            except Exception as e:
                print(f"⛔:{e}")

        self.addSpeed(10)

    def addSpeed(self, velocidad):
        self.velocidad += velocidad
        return self.velocidad

    def downSpeed(self, velocidad):
        self.velocidad -= velocidad
        return self.velocidad

    def currentSpeed(self):
        print(f"Current Speed 🛫: {self.velocidad}")

    def landing(self):
        for i in range(5):
            print("🛬🛬 Landing in Progress 🚷🚷")

            try:
                time.sleep(1 - (i * 50))
            except Exception as e:
                print(f"⛔:{e}")

    def descent(self):
        speedUpString = ""

        for i in range(self.velocidad):
            speedUpString += "🛬"

        for i in range(10):
            speedUpString = speedUpString[:-1]
            print(f"🛬🛬Descent in Progress🛬🛬: {speedUpString}")

            try:
                time.sleep(1 - (self.velocidad + i * 50))
            except Exception as e:
                print(f"⛔:{e}")

        self.downSpeed(10)

    # Getters and Setters
    def getMarca(self):
        return self.marca

    def setMarca(self, marca):
        self.marca = marca

    def getTipo(self):
        return self.tipo

    def setTipo(self, tipo):
        self.tipo = tipo

    def getVelocidad(self):
        return self.velocidad

    def setVelocidad(self, velocidad):
        self.velocidad = velocidad

    def chargeFuel(self, litros):
        print(f"⛽ Charging Fuel: {litros} 💧💧💧")